codec/reflectcodec/type_codec.go

```diff
@@ -31,100 +24,63 @@ var (
 	errExtraSpace   = errors.New("trailing buffer space")
 )
 
-// Codec handles marshaling and unmarshaling of structs
-type codec struct {
-	lock         sync.RWMutex
-	tagName      string
-	maxSliceLen  int
-	nextTypeID   uint32
-	typeIDToType map[uint32]reflect.Type
-	typeToTypeID map[reflect.Type]uint32
+type TypeCodec interface {
+	// UnpackPrefix unpacks the prefix of an interface from the given packer.
+	// The prefix specifies the concrete type that the interface should be
+	// deserialized into. This function returns a new instance of that concrete
+	// type. The concrete type must implement the given type.
+	UnpackPrefix(*wrappers.Packer, reflect.Type) (reflect.Value, error)
 
-	fieldLock sync.Mutex
-	// Key: a struct type
-	// Value: Slice where each element is index in the struct type
-	// of a field that is serialized/deserialized
-	// e.g. Foo --> [1,5,8] means Foo.Field(1), etc. are to be serialized/deserialized
-	// We assume this cache is pretty small (a few hundred keys at most)
-	// and doesn't take up much memory
-	serializedFieldIndices map[reflect.Type][]int
+	// PackPrefix packs the prefix for the given type into the given packer.
+	// This identifies the bytes that follow, which are the byte representation
+	// of an interface, as having the given concrete type.
+	// When deserializing the bytes, the prefix specifies which concrete type
+	// to deserialize into.
+	PackPrefix(*wrappers.Packer, reflect.Type) error
 }
 
-// Codec marshals and unmarshals
-type Codec interface {
-	Registry
-	MarshalInto(interface{}, *wrappers.Packer) error
-	Unmarshal([]byte, interface{}) error
+// genericCodec handles marshaling and unmarshaling of structs with a generic
+// implementation for interface encoding.
+//
+// A few notes:
+// 1) We use "marshal" and "serialize" interchangeably, and "unmarshal" and "deserialize" interchangeably
+// 2) To include a field of a struct in the serialized form, add the tag `{tagName}:"true"` to it. `{tagName}` defaults to `serialize`.
+// 3) These typed members of a struct may be serialized:
+//    bool, string, uint[8,16,32,64], int[8,16,32,64],
+//	  structs, slices, arrays, interface.
+//	  structs, slices and arrays can only be serialized if their constituent values can be.
+// 4) To marshal an interface, you must pass a pointer to the value
+// 5) To unmarshal an interface,  you must call codec.RegisterType([instance of the type that fulfills the interface]).
+// 6) Serialized fields must be exported
+// 7) nil slices are marshaled as empty slices
+type genericCodec struct {
```

Comment:
Can we add a comment that this struct isn't safe for concurrent use?

Reply:
What's not safe about it?

Reply:
nvm, I forgot we moved the lock to the "struct fielder"